### PR TITLE
chore(web): updates TypeScript version to 5.4.5

### LINKED ISSUE #11375: chore(common): update Typescript to 5.x

### PR DESCRIPTION
**Pull request by jahorton**
Fixes #11375.

Also updates @types/node to a consistent version across all packages, as it caused a conflict after the TS upgrade.  I'd left it with partly 18.0 and partly 20.0 in #11319, and it looks like our hand is forced with the TS upgrade.

@keymanapp-test-bot skip

### CODE REVIEW

**Comment by jahorton** [comment]
Current build failure is from commit 1, which really should be part of the prior PR but wasn't caught at the time.

**Review by ermshiperete — APPROVED** [review]

File: common/tools/hextobin/package.json [hunk 1]
```diff
@@ -15,8 +15,5 @@
   "main": "build/index.js",
   "bin": {
     "hextobin": "build/hextobin.js"
-  },
-  "devDependencies": {
-    "@types/node": "^18.7.18"
   }
```
Comment (mcdurdin) [inline common/tools/hextobin/package.json:18]: I wish there was a way in package.json to clearly document dependencies for a package while hoisting the version requirement to the top level. It's useful documentation to have :sob:

File: tsconfig.base.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "compilerOptions": {
-    "module": "ES2022",
+    "module": "node16",
```
Comment (mcdurdin) [inline tsconfig.base.json:3]: This is a scary config change! Are you confident that it isn't going to break bundling, etc, for kmc, developer/server, etc?

File: web/tsconfig.base.json [hunk 3]
```diff
@@ -8,7 +8,7 @@
     "allowSyntheticDefaultImports": true,
 
     "lib": ["es6"],
-    "module": "es6",
+    "module": "node16",
```
Comment (mcdurdin) [inline web/tsconfig.base.json:11]: Can this be deleted?

File: tsconfig.base.json (hunk 2 above)
Comment (jahorton) [reply tsconfig.base.json:3]: Yes.  This is the prescribed "correct" way to do things in TS 5; we _cannot_ leave `ES2022` in place, as it's now a TS config error to do so.  Use of `"moduleResolution": "node16"` explicitly requires use of `"module": "node16"`.  They _must_ match.  We cannot update to TS 5.0+ without doing this.

Would you like me to add a user test that verifies that the resulting Developer build is unaffected?  If so, would you prefer it here, on the final related PR (#11464), or both?

Comment (mcdurdin) [reply tsconfig.base.json:3]: Okay, that makes sense, thanks.

It'd be good for the user test to run on the final PR in the chain. It _should_ be fine, as far as I can tell, but I am getting increasingly wary with config changes and unanticipated outcomes in js/ts world, particularly where we have so many modules with varying histories.

I think we need user tests of the compile toolchain (compiling within Keyman Developer IDE is fine) and Keyman Developer Server (so testing web keyboards in Keyman Developer IDE).

Comment (jahorton) [reply tsconfig.base.json:3]: FWIW, the Developer build from this PR (which passed) seems to have a fully-functioning Web test-host page when I test it locally myself.  Recompilation of the keyboard (after a .keyman-touch-layout edit) seems to proceed fine.

That said... I noticed something really strange.  If I edited the touch-layout a bit, then built the keyboard and tested via Developer Server, all was fine.  If I then returned to the touch-layout editor and used `CTRL-Z` (undo) once, saved, then rebuilt the keyboard and resumed testing... the results did _not_ get updated.  I was able to determine that the touch-layout (undo) edits did not make it into the compiled .js in this case.

Re-editing the touch layout, then copying-and-pasting the original value for edited parts of the touch layout back in place, then doing the usual save -> rebuild -> test, acted as expected, updating the test-host keyboard properly once again.

Comment (mcdurdin) [reply tsconfig.base.json:3]: > Looks to be something undo-specific?

Possibly yes.

Comment (jahorton) [reply tsconfig.base.json:3]: Forgot to link it previously:  I spun that "something really strange" observation off into #11514.

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.45-alpha](https://keyman.com/downloads/releases/alpha/18.0.45)

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.45-alpha](https://keyman.com/downloads/releases/alpha/18.0.45)

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.45-alpha](https://keyman.com/downloads/releases/alpha/18.0.45)

**Comment by keyman-server** [comment]
Changes in this pull request will be available for download in [Keyman version 18.0.45-alpha](https://keyman.com/downloads/releases/alpha/18.0.45)

**Pull request merged**